feat(trends): Cleaning up the trends endpoint and adding support for filtering trend percentage and difference

wmak opened the pull request:
- This adds support for querying on aliases on the more complicated
  trends functions.
  - eg. percentage(aggregate_range_2, aggregate_range_1):>1 becomes
  	trend_percentage():>0%
  - Note that this is 0% since we subtract by 1 when we display the
    trend percentage

wmak committed:
ref: Refactoring the trends endpoint

- This adds support for querying on aliases on the more complicated
  trends functions.
  - eg. percentage(aggregate_range_2, aggregate_range_1):>1 becomes
  	trend_percentage():>0%
  - Note that this is 0% since we subtract by 1 when we display the
    trend percentage

wmak committed:
fix: updating aliases in tests to match the new ones

wmak committed:
ref: Adding tests

- Also fixing some mistakes in how the aliases are constructed (yay
  tests)

wmak committed:
ref: Renaming improvement to improved

wmak committed:
ref: Introduce new alias fields to the frontend

wmak committed:
ref: Renaming the aggregate ranges as well

- This way the baseline aggregates will always be `aggregate_range_x`

wmak committed:
fix: some places where params is not defined

wmak committed:
fix: Some more broken tests from earlier refactors

Zylphrex commented on src/sentry/api/event_search.py:
```diff
@@ -2154,17 +2173,22 @@ def result_type_fn(function_arguments, parameter_values):
+                FunctionArg("denominator"),
+                FunctionArg("query_alias"),
+            ],
+            # aggregate and not a column since it needs to act on aggregates itself
```
I don't think this is the reason why `percentage` is defined as an aggregate here. I believe the reason is that you need to pass the expression through to clickhouse as is without modifications. And defining it as a column does not let you do that at this time.

wmak replied:
So during exploration of this (and a few other solutions) I tried flipping this to a column, which then meant that conditions on it ended up in the `WHERE` which would break since the conditions that we're using this with contain aggregates, and have to end up in the `HAVING`

Zylphrex commented on src/sentry/static/sentry/app/views/performance/trends/changedProjects.tsx:
```diff
@@ -26,13 +26,7 @@ import {
-  getTrendProjectId,
-  getCurrentTrendFunction,
-} from './utils';
+import {modifyTrendView, normalizeTrends, trendToColor, getTrendProjectId} from './utils';
```
This whole file is being removed in #21753.

Zylphrex reviewed (commented):
Nit: Can this PR be renamed because this is definitely not a refactor.

markstory commented on src/sentry/api/endpoints/organization_events_trends.py:
```diff
@@ -14,64 +14,150 @@
+                lambda aggregate_filter: [
+                    "trend_percentage",
+                    NEGATION_MAP[aggregate_filter.operator]
+                    if trend_type == "improved"
```
Should improved/regressed be constants?

markstory commented on src/sentry/api/event_search.py:
```diff
@@ -171,6 +172,7 @@ def translate(pat):
 alt_date_format      = ~r"\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}(\.\d{1,6})?(\+\d{2}:\d{2})?)?(?=\s|\)|$)"
 rel_date_format      = ~r"[\+\-][0-9]+[wdhm](?=\s|\)|$)"
 duration_format      = ~r"([0-9\.]+)(ms|s|min|m|hr|h|day|d|wk|w)(?=\s|\)|$)"
+percentage_format    = ~r"([0-9\.]+)%"
```
This will allow `0.0.0.0`. I don't think that's what you want. The pattern of `(?:[0-9]+)?\.(?:[0-9]+)` would prevent that.

wmak replied:
🙃 oh nice `duration_format` will allow `0.0.0.0.0ms` too I'll update that as well

wmak replied:
_annnnd_ `numeric_value` (which is what I based `duration_format` on) 🙃 🙃 🙃

wmak replied:
Okay, going to keep this as is, the error message of `*:5.5.5.5.5%` becomes `5.5.5.5 is not a valid percentage value`. Improving this regex results in the query being parsed as a message instead.

wmak commented on src/sentry/api/endpoints/organization_events_trends.py:
```diff
@@ -14,64 +14,150 @@
+
+
+# This is to flip conditions beteween trend types
+NEGATION_MAP = {
```
I don't mind `CORRESPONDENCE_MAP` :+1:

markstory commented on src/sentry/static/sentry/app/views/performance/trends/types.ts:
```diff
@@ -80,28 +82,24 @@ type BaseTrendsTransaction = {
-  percentage_count_range_2_count_range_1: number;
+  trend_percentage: number;
+  trend_difference: number;
+  count_percentage: number;
```
These field names are much nicer :clap:

wmak committed:
ref: Addressing PR comments

- Negation -> correspondence
- consistency in the trend_column aliasing
- improved/regression to constants
- Updating comment on why percentage is an aggregate

wmak committed:
Merge branch 'master' into ref/trends-endpoint

wmak committed:
fix: missing underscore for minus

wmak committed:
fix: Removed TrendsTransaction accidentally
